state/applicationoffers.go

```diff
@@ -489,6 +506,59 @@ func (s *applicationOffers) filterOffers(
 	return out, nil
 }
 
+func (s *applicationOffers) filterOffersByConnectedUser(
+	in []applicationOfferDoc,
+	users []string,
+) ([]applicationOfferDoc, error) {
+
+	if len(users) == 0 {
+		return in, nil
+	}
+
+	offerUUIDS := make(set.Strings)
+	for _, username := range users {
+		conns, err := s.st.OfferConnectionsForUser(username)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		for _, oc := range conns {
+			offerUUIDS.Add(oc.OfferUUID())
+		}
+	}
+	var out []applicationOfferDoc
+	for _, doc := range in {
+		if offerUUIDS.Contains(doc.OfferUUID) {
+			out = append(out, doc)
+		}
+	}
+	return out, nil
+}
+
+func (s *applicationOffers) filterOffersByAllowedUser(
+	in []applicationOfferDoc,
+	users []string,
+) ([]applicationOfferDoc, error) {
+
+	if len(users) == 0 {
+		return in, nil
+	}
+
+	var out []applicationOfferDoc
+	for _, doc := range in {
+		offerUsers, err := s.st.GetOfferUsers(doc.OfferUUID)
```

Comment:
Is this going to scale poorly if most offers are public?
Would it be better to only load the UUIDs that we haven't filtered so far and check if the user is allowed to access them?

Reply:
For Juju 2.3, all offers will be controller via ACLs. GetOfferUsers() is an existing API call that is being re-used here. We need to adjust things anyway when the JAAS model is supported.

Reply:
because this is the last filter step done, the overall offers list has already been filtered down using the other 2 filters